distributed/shuffle/tests/test_shuffle.py

```diff
@@ -2165,3 +2166,23 @@ async def test_set_index_p2p(c, s, *workers):
     await c.close()
     await asyncio.gather(*[check_worker_cleanup(w) for w in workers])
     await check_scheduler_cleanup(s)
+
+
+@gen_cluster(
+    client=True,
+)
+async def test_p2p_shuffle_timeout(c, s, a, b):
+    def block():
+        time.sleep(5)
+
+    df = pd.DataFrame({"a": [1, 2, 3, 4, 5, 6, 7, 8], "b": 1})
+    ddf = dd.from_pandas(df, npartitions=3)
+    ddf = ddf.set_index("a", shuffle="p2p", divisions=(1, 3, 8))
+    assert ddf.npartitions == 2
+
+    with dask.config.set({"distributed.comm.timeouts.connect": "3s"}):
+        await asyncio.gather(c.run(block, workers=[a.address]), c.compute(ddf))
```

Comment:
I would've expected this to fail with a sufficiently large value for `distributed.comm.timeouts.connect`, but it seems to pass regardless of what this is set to - not sure if I've done something wrong here in terms of blocking the event loop on a single worker

Reply:
I'll take a look!

Reply:
Sorry, I haven't been had the time to look at this as I was busy hotfixing a P2P regression. I'll be out for the next three weeks, but maybe @crusaderky or @fjetter can help you out here. One thing to look out for is the correct timing of the event loop blockage. If it's blocked too early or too late it won't interfere with the connection and things will just work.

Reply:
One way to tackle this would be through mocking (though it's not my favorite since won't account for changes to the mocked code):
```python3

class FlakyConnectionPool(ConnectionPool):
    def __init__(self, *args, failing_connects=0, **kwargs):
        self.cnn_count = 0
        self.failing_connects = failing_connects
        super().__init__(*args, **kwargs)

    async def connect(self, *args, **kwargs):
        self.cnn_count += 1
        if self.cnn_count > self.failing_connects:
            return await super().connect(*args, **kwargs)
        else:
            raise OSError()

@gen_cluster(
    client=True,
    config={"distributed.comm.retry.count": 0, "distributed.p2p.comm.retry.count": 0}
)
async def test_p2p_flaky_connect_fails_without_retry(c, s, a, b):

    df = dask.datasets.timeseries(
        start="2000-01-01",
        end="2000-01-10",
        dtypes={"x": float, "y": float},
        freq="10 s",
    )
    x = dd.shuffle.shuffle(df, "x", shuffle="p2p")
    
    rpc = await FlakyConnectionPool(failing_connects=1)
    
    with mock.patch.object(a, "rpc", rpc):
        with pytest.raises(RuntimeError, match="P2P shuffling.*transfer"):
            await c.compute(x)
    
    await check_worker_cleanup(a)
    await check_worker_cleanup(b)
    await c.close()
    await check_scheduler_cleanup(s)


@gen_cluster(
    client=True,
    config={"distributed.comm.retry.count": 0, "distributed.p2p.comm.retry.count": 1}
)
async def test_p2p_flaky_connect_recover_with_retry(c, s, a, b):

    df = dask.datasets.timeseries(
        start="2000-01-01",
        end="2000-01-10",
        dtypes={"x": float, "y": float},
        freq="10 s",
    )
    x = dd.shuffle.shuffle(df, "x", shuffle="p2p")
    
    rpc = await FlakyConnectionPool(failing_connects=1)
    
    with mock.patch.object(a, "rpc", rpc):
        await c.compute(x)
    
    await check_worker_cleanup(a)
    await check_worker_cleanup(b)
    await check_scheduler_cleanup(s)
```

Reply:
This version is better. It's still mocking, but it triggers an actual timeout: 
```python3

class FlakyConnectionPool(ConnectionPool):
    def __init__(self, *args, failing_connects=0, **kwargs):
        self.failed_attempts = 0
        self.failing_connects = failing_connects
        super().__init__(*args, **kwargs)

    async def connect(self, *args, **kwargs):
        if self.failed_attempts >= self.failing_connects:
            return await super().connect(*args, **kwargs)
        
        with dask.config.set({"distributed.comm.timeouts.connect": "0 ms"}):
            try:
                _ = await super().connect(*args, **kwargs)
            except Exception as e:
                self.failed_attempts += 1
                raise e
            else:
                raise RuntimeError("This should not happen, connecting should fail!")
    

@gen_cluster(
    client=True,
    config={"distributed.comm.retry.count": 0, "distributed.p2p.comm.retry.count": 0}
)
async def test_p2p_flaky_connect_fails_without_retry(c, s, a, b):

    df = dask.datasets.timeseries(
        start="2000-01-01",
        end="2000-01-10",
        dtypes={"x": float, "y": float},
        freq="10 s",
    )
    x = dd.shuffle.shuffle(df, "x", shuffle="p2p")
    
    rpc = await FlakyConnectionPool(failing_connects=1)
    
    with mock.patch.object(a, "rpc", rpc):
        with raises_with_cause(expected_exception=RuntimeError, match="P2P shuffling.*transfer", expected_cause=OSError, match_cause=None):
            await c.compute(x)
    
    await check_worker_cleanup(a)
    await check_worker_cleanup(b)
    await c.close()
    await check_scheduler_cleanup(s)


@gen_cluster(
    client=True,
    config={"distributed.comm.retry.count": 0, "distributed.p2p.comm.retry.count": 1}
)
async def test_p2p_flaky_connect_recover_with_retry(c, s, a, b):

    df = dask.datasets.timeseries(
        start="2000-01-01",
        end="2000-01-10",
        dtypes={"x": float, "y": float},
        freq="10 s",
    )
    x = dd.shuffle.shuffle(df, "x", shuffle="p2p")
    
    rpc = await FlakyConnectionPool(failing_connects=1)
    
    with mock.patch.object(a, "rpc", rpc):
        await c.compute(x)
    
    await check_worker_cleanup(a)
    await check_worker_cleanup(b)
    await check_scheduler_cleanup(s)
```